pkg/soc/soc.go

```diff
@@ -0,0 +1,234 @@
+	publicKey, err := signer.PublicKey()
+	if err != nil {
+		return err
+	}
+	ownerAddressBytes, err := crypto.NewEthereumAddress(*publicKey)
+	if err != nil {
+		return err
+	}
+	ownerAddress, err := NewOwner(ownerAddressBytes)
+	if err != nil {
+		return err
+	}
+	s.signer = signer
+	s.owner = ownerAddress
+	return nil
+}
+
+// OwnerAddress returns the ethereum address of the signer of the Chunk.
+func (s *Soc) OwnerAddress() []byte {
+	return s.owner.address
+}
+
+// Address returns the soc Chunk address.
+func (s *Soc) Address() (swarm.Address, error) {
+	return CreateAddress(s.id, s.owner)
+}
+
+// FromChunk recreates an Soc representation from swarm.Chunk data.
+func FromChunk(sch swarm.Chunk) (*Soc, error) {
+	chunkData := sch.Data()
+	if len(chunkData) < minChunkSize {
+		return nil, errors.New("less than minimum length")
+	}
+
+	// add all the data fields
+	s := &Soc{}
+	cursor := 0
+
+	s.id = chunkData[cursor : cursor+IdSize]
+	cursor += IdSize
+
+	s.signature = chunkData[cursor : cursor+SignatureSize]
+	cursor += SignatureSize
+
+	spanBytes := chunkData[cursor : cursor+swarm.SpanSize]
+	cursor += swarm.SpanSize
+
+	ch, err := content.NewContentChunkWithSpanBytes(chunkData[cursor:], spanBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	toSignBytes, err := toSignDigest(s.id, ch.Address().Bytes())
+	if err != nil {
+		return nil, err
+	}
+
+	// recover owner information
+	recoveredEthereumAddress, err := recoverAddress(s.signature, toSignBytes)
+	if err != nil {
+		return nil, err
+	}
+	owner, err := NewOwner(recoveredEthereumAddress)
+	if err != nil {
+		return nil, err
+	}
+	s.owner = owner
+	s.Chunk = ch
+
+	return s, nil
+}
+
+// ToChunk generates a signed chunk payload ready for submission to the swarm network.
+//
+// The method will fail if no signer has been added.
+func (s *Soc) ToChunk() (swarm.Chunk, error) {
+	var err error
+	if s.signer == nil {
+		return nil, errors.New("signer missing")
+	}
+
+	// generate the data to sign
+	toSignBytes, err := toSignDigest(s.id, s.Chunk.Address().Bytes())
+	if err != nil {
+		return nil, err
+	}
+
+	// sign the chunk
+	signature, err := s.signer.Sign(toSignBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	// prepare the payload
+	buf := bytes.NewBuffer(nil)
+	buf.Write(s.id)
+	buf.Write(signature)
+	buf.Write(s.Chunk.Data())
+
+	// create chunk
+	socAddress, err := s.Address()
+	if err != nil {
+		return nil, err
+	}
+	return swarm.NewChunk(socAddress, buf.Bytes()), nil
+}
+
+// toSignDigest creates a digest suitable for signing to represent the soc.
+func toSignDigest(id Id, sum []byte) ([]byte, error) {
+	h := swarm.NewHasher()
+	_, err := h.Write(id)
+	if err != nil {
+		return nil, err
+	}
+	_, err = h.Write(sum)
+	if err != nil {
+		return nil, err
+	}
+	return h.Sum(nil), nil
+}
+
+// CreateAddress creates a new soc address from the soc id and the ethereum address of the signer
+func CreateAddress(id Id, owner *Owner) (swarm.Address, error) {
```

Comment:
`NewAddress` might be more idiomatic.

Reply:
I am not sure. `New` suggests - to me at least - instantiate an object, reserve memory etc. This is more like a helper function.

Reply:
Makes sense!